junit-platform-commons/junit-platform-commons.gradle.kts

```diff
@@ -11,11 +13,23 @@ dependencies {
 	api(platform(projects.junitBom))
 
 	compileOnlyApi(libs.apiguardian)
+
+	osgiVerification(projects.junitJupiterEngine)
+	osgiVerification(projects.junitPlatformLauncher)
 }
 
 tasks.jar {
 	val release9ClassesDir = sourceSets.mainRelease9.get().output.classesDirs.singleFile
 	inputs.dir(release9ClassesDir).withPathSensitivity(PathSensitivity.RELATIVE)
+
+	withConvention(BundleTaskConvention::class) {
+		bnd("""
+			Require-Capability:\
+				org.junit.platform.engine;\
+					filter:='(&(org.junit.platform.engine=*)(version>=${'$'}{version_cleanup;${rootProject.property("version")!!}})(!(version>=${'$'}{versionmask;+;${'$'}{version_cleanup;${rootProject.property("version")!!}}})))'
```

Comment:
I don't think that this is the way to go, for a couple of reasons:
1. `commons` is a utility library. It doesn't require an engine to be present in order to provide the functionality that it is designed to provide. So this doesn't seem right the right place to put this. If anything, this requirement belongs on the launcher bundle, but even there I think it is out of place (by way of comparison, `org.apache.felix.scr` does not have a `Require-Capability: osgi.service;filter:='(&(objectClass=*)` directive to wildcard include some stuff).
2. I'm not quite sure what the resolver does with the filter "*" - will it pull in exactly one match (indiscriminately), or all matches? Neither of these seem correct:
   1. If it's pulling in all of them, then it may pull in engines that aren't needed.
   2. If it's only pulling in one, then there is no guarantee that it is the right one.

In the case, where you've only got the Jupiter engine and Jupiter tests in your repo, it will all work by default. But if you've got mixed tests in your workspace you might not get the right bundle assembly. You might even end up with a set of `-runbundles` that resolves at runtime but which silently skips a bunch of tests (maybe even all of the tests!) because the engine those tests require has not been included in `-runbundles`.

Which brings us back to the fundamental point (which this PR is designed to address): the question "which engines are required?" is a question that applies to tests. If it is a JUnit 3/4 test, then the test bundle requires the Vintage engine; if it a Jupiter test then it requires the Jupiter engine. As these requirements only make sense when considered with reference to a particular test, then the most natural place to declare them is in the manifest of the bundle that contains the test (ie, the subject of bndtools/bnd#4719).

Reply:
> commons is a utility library. It doesn't require an engine to be present in order to provide the functionality that it is designed to provide. So this doesn't seem right the right place to put this. If anything, this requirement belongs on the launcher bundle, but even there I think it is out of place (by way of comparison, org.apache.felix.scr does not have a Require-Capability: osgi.service;filter:='(&(objectClass=*) directive to wildcard include some stuff).

The point of it was to make every client lib dependent on having an engine at runtime. I don't think there is a valid argument against that since all client APIs have a hard dependency on commons. Therefore if you make the requirement on commons no matter how many more client APIs you create they will always result in requiring an engine at runtime.

> I'm not quite sure what the resolver does with the filter "*" - will it pull in exactly one match (indiscriminately), or all matches? Neither of these seem correct:
>
>    If it's pulling in all of them, then it may pull in engines that aren't needed.
>    If it's only pulling in one, then there is no guarantee that it is the right one.

It "requires" at least one to exist, which one is dependent on what's available in the resolution repos. I've never used both together so if that's possible, then I guess you're right that we can make each one targetable, but there is no need to change this wildcard. We can pair this concept with a per engine requirement which the dev can use. Let me add that capability.

> Which brings us back to the fundamental point (which this PR is designed to address): the question "which engines are required?" is a question that applies to tests. If it is a JUnit 3/4 test, then the test bundle requires the Vintage engine; if it a Jupiter test then it requires the Jupiter engine.

This I can easily add. Stay tuned...

Reply:
Actually, you don't even have to add anything to the existing.

All a test developer has to do to add both engines is to write the following in the bnd file of their test bundle (bundle which contains the `Test-Cases` header):

```
Require-Capability: \
    org.junit.platform.engine;filter:='(org.junit.platform.engine=junit-jupiter)',\
    org.junit.platform.engine;filter:='(org.junit.platform.engine=junit-vintage)'
```
OR if they prefer the configuration-as-code approach:
```
@org.osgi.annotation.bundle.Requirement(
    namespace = "org.junit.platform.engine", name = "junit-jupiter")
@org.osgi.annotation.bundle.Requirement(
    namespace = "org.junit.platform.engine", name = "junit-vintage")
```
the resolve process will take care of the rest.

Reply:
@kriegfrj I do now see your point about the model with commons however. since vintage engine depends on it. Making some tweaks to the model.

Reply:
> Actually, you don't even have to add anything to the existing.
> 
> All a test developer has to do to add both engines is to write the following in the bnd file of their test bundle (bundle which contains the `Test-Cases` header):
> 
> *snip*
>
> the resolve process will take care of the rest.

Correct! That's exactly what is needed. (Btw, the solution proposed in bndtools/bnd#4719 is for Bnd to do exactly this on behalf of the developer to relieve them of the burden of needing to do it manually.)

These Req-Caps make the Req-Cap on commons redundant though. At best, the requirement on commons will pull in the same engine; at worst, it will pull in a different one that's not actually required.

Reply:
@kriegfrj I'm going to resolve this convo for now in favour of going over the updates :)

Reply:
> Btw, the solution proposed in bndtools/bnd#4719 is for Bnd to do exactly this on behalf of the developer to relieve them of the burden of needing to do it manually.

First off, with this PR in the case of jupiter tests there's nothing for the dev or bnd to do. It's already working transparently. Add dependencies -> write tests -> resolve -> run tests.

Now, the hard case is JUnit 3&4 where you have two possible runtimes from bnd's point of view, the legacy tester or junit-platform. From the test developer point of view they are still just writing tests. They then have to have opted-into one of the two runtimes. Now in the case of opting to use junit-platform I don't see it as being possible for bnd to optimize away the decision to pull in the vintage engine. I just think that the one extra line for the proper requirement fits nicely beside the existing opt-in as pretty low friction since it's done exactly once. Furthermore, the frequency that we'll see both junit-platform AND the vintage engine used together is no worth optimizing that even further in my opinion.